### PR TITLE
Logic feature

### PR DESCRIPTION
**Pull request by arturAR**
Implementing messages in the state machine.

### CODE REVIEW

File: src/main/java/com/java_academy/ClientApp.java [hunk 1]
```diff
@@ -21,18 +26,26 @@ public static void main(String[] args) {
 
         Socket socket = new Socket();
```
Comment (mprtcz) [inline src/main/java/com/java_academy/ClientApp.java:28]: Those empty lines, why?

File: src/main/java/com/java_academy/ClientApp.java [hunk 2]
```diff
@@ -21,18 +26,26 @@ public static void main(String[] args) {
...
 
-        SocketProvider socketProvider = new ClientSocketProvider(socket, messageSupplier -> {
-            System.out.println("message from the server: " + messageSupplier.get());
-
-        });
+        SocketProvider socketProvider = new ClientSocketProvider(socket);
 
         Connector connector = new Connector(socketProvider);
 
+        connector.addMessageReseiverListenerToSocketProvider(messageSupplier -> System.out.println("message from the server: " + messageSupplier.get()));
```
Comment (mprtcz) [inline src/main/java/com/java_academy/ClientApp.java:33]: addMessageRe**c**eiverListenerToSocketProvider, but still rather a confusing name

File: src/main/java/com/java_academy/ClientApp.java [hunk 3]
```diff
@@ -21,18 +26,26 @@ public static void main(String[] args) {
...
+
         InetSocketAddress inetSocketAddress = new InetSocketAddress("localhost", 3000);
         if (connector.connect(inetSocketAddress)){
             String message = "HELLO from Client";
             System.out.println("Sending message \"" + message + "\" to the server");
             connector.sendMessage(new MessageObject(Players.FIRST_PLAYER, message));
+
+            Connector.getExecutor().schedule(new Runnable() {
```
Comment (mprtcz) [inline src/main/java/com/java_academy/ClientApp.java:41]: Fancy stuff

File: src/main/java/com/java_academy/ClientApp.java (hunk 2 above)
Comment (mprtcz) [inline src/main/java/com/java_academy/ClientApp.java:33]: Maybe some logger instead of sout?

File: src/main/java/com/java_academy/ServerApp.java [hunk 4]
```diff
@@ -21,13 +26,21 @@
     public static void main(String[] args) {
         try {
             ServerSocket serverSocket = new ServerSocket();
-            SocketProvider socketProvider = new ServerSocketProvider(serverSocket, messageSupplier -> {
-                System.out.println("message from a Client: " + messageSupplier.get());
-            });
+            SocketProvider socketProvider = new ServerSocketProvider(serverSocket);
 
             InetSocketAddress inetSocketAddress = new InetSocketAddress("localhost", 3000);
```
Comment (mprtcz) [inline src/main/java/com/java_academy/ServerApp.java:31]: I know it's not added change but you'd rather want those to be static constants.

File: src/main/java/com/java_academy/ServerApp.java [hunk 5]
```diff
@@ -21,13 +26,21 @@
...
 
             Connector connector = new Connector(socketProvider);
+
+            connector.addMessageReseiverListenerToSocketProvider(messageSupplier -> {
+                System.out.println("message from a Client: " + messageSupplier.get());
+                if (messageSupplier.get().equals(CLOSE_MESSAGE)){
+                    connector.sendMessage(new MessageObject(Players.FIRST_PLAYER, CLOSE_MESSAGE));
+                    connector.sendMessage(new MessageObject(Players.SECOND_PLAYER, CLOSE_MESSAGE));
```
Comment (mprtcz) [inline src/main/java/com/java_academy/ServerApp.java:39]: 👍

File: src/main/java/com/java_academy/logic/state_machine/core/GameState.java [hunk 6]
```diff
@@ -9,10 +9,14 @@
...
- * isEndingState - check state
+ * isEndingState - check state, default is not end state
  * */
 public interface GameState {
+	
     void display(Consumer<MessageObject> displayConsumer);
     GameState changeState(String message);
-    boolean isEndingState();
+    
+    default boolean isEndingState() {
```
Comment (ghost) [inline src/main/java/com/java_academy/logic/state_machine/core/GameState.java:19]: :+1:

File: src/main/java/com/java_academy/logic/state_machine/BattleResultState.java [hunk 7]
```diff
@@ -19,17 +21,13 @@ public BattleResultState(Players winner) {
 
     @Override
     public void display(Consumer<MessageObject> outputConsumer) {
-        outputConsumer.accept(new MessageObject(winner, "YOU WIN!"));
-        outputConsumer.accept(new MessageObject(winner.getOpponent(), "YOU LOST!"));
+    	outputConsumer.accept(new MessageObject(winner, MessageCreator.createJsonMessageByKey("you.win")));
```
Comment (mprtcz) [inline src/main/java/com/java_academy/logic/state_machine/BattleResultState.java:24]: Nice method name 👍

File: src/main/resources/Messages_pl_PL.properties [hunk 8]
```diff
@@ -9,4 +9,11 @@ TRYING_TO_TERMINATE_EXECUTOR =
...
-CANT_SEND_MESSAGE_TO_SERVER =
\ No newline at end of file
+CANT_SEND_MESSAGE_TO_SERVER =
+new.game = NOWA GRA!
+set.fleet = Ustawam statki losowo!
+your.turn = Twoja kolej! Strzelaj!
+not.your.turn = Czekaj, teraz strzela przeciwnik!
+you.win = Wygra\u0142e\u015B! Brawo!
+you.lose = Przegra\u0142e\u015B!
+end.game = KONIEC GRY!
```
Comment (ghost) [inline src/main/resources/Messages_pl_PL.properties:19]: :+1:

**Review by ghost — APPROVED** [review]

File: src/main/java/com/java_academy/logic/state_machine/PlayerActionState.java [hunk 9]
```diff
@@ -22,30 +24,33 @@ public PlayerActionState(Players currentPlayer){
     }
     @Override
     public void display(Consumer<MessageObject> displayConsumer) {
-        displayConsumer.accept(new MessageObject(Players.FIRST_PLAYER, "SHOT"));
-        displayConsumer.accept(new MessageObject(Players.SECOND_PLAYER, "WAIT"));
-    }
+    	displayConsumer.accept(new MessageObject(currentPlayer, MessageCreator.createJsonMessageByKey("your.turn")));
+        displayConsumer.accept(new MessageObject(currentPlayer.getOpponent(), MessageCreator.createJsonMessageByKey("not.your.turn")));
```
Comment (mprtcz) [inline src/main/java/com/java_academy/logic/state_machine/PlayerActionState.java:28]: 👍

File: src/main/java/com/java_academy/logic/state_machine/PlayerActionState.java [hunk 10]
```diff
@@ -22,30 +24,33 @@ public PlayerActionState(Players currentPlayer){
...
         }
 
-        return new PlayerEndActionState(currentPlayer);
+        return new PlayerEndActionState(currentPlayer, markedIndexes);
+    }
+    
+    private MarkedIndexes dropNuke(String inputMessage) {
+    	Integer index = Integer.parseInt(inputMessage);
+    	Attack attack = new NukeAttack(currentPlayer.getOpponent().getPlayer().getBoard());
+    	return attack.attack(index);
```
Comment (mprtcz) [inline src/main/java/com/java_academy/logic/state_machine/PlayerActionState.java:47]: A lot of attacks here

File: src/main/java/com/java_academy/logic/state_machine/PlayerActionState.java [hunk 11]
```diff
@@ -22,30 +24,33 @@ public PlayerActionState(Players currentPlayer){
...
     }
+    
+    private MarkedIndexes shootOnField(String inputMessage) {
+    	Integer index = Integer.parseInt(inputMessage);
+    	Attack attack = new NormalAttack(currentPlayer.getOpponent().getPlayer().getBoard() );
```
Comment (mprtcz) [inline src/main/java/com/java_academy/logic/state_machine/PlayerActionState.java:52]: Ugly chained getters, come on

File: src/main/java/com/java_academy/logic/state_machine/PlayerEndActionState.java [hunk 12]
```diff
@@ -14,19 +18,24 @@
...
+    	markedIndexes.setIsMyBoard(true);
+    	displayConsumer.accept(new MessageObject(currentPlayer, MessageCreator.createJsonMarkedIndexes(markedIndexes)));
+		markedIndexes.setIsMyBoard(false);
+		displayConsumer.accept(new MessageObject(currentPlayer.getOpponent(), MessageCreator.createJsonMarkedIndexes(markedIndexes)));
     }
 
     @Override
     public GameState changeState(String message) {
-        hasBeenHit = somethingWasHitted();
+        hasBeenHit = somethingWasHit(markedIndexes.getMap());
```
Comment (mprtcz) [inline src/main/java/com/java_academy/logic/state_machine/PlayerEndActionState.java:38]: 👍

File: src/main/java/com/java_academy/logic/state_machine/PlayerEndActionState.java [hunk 13]
```diff
@@ -38,16 +47,16 @@ public GameState changeState(String message) {
...
-        return false;
-    }
-
     private boolean checkIfPlayerWon(Players player) {
         return player.getOpponent().getPlayer().hasNoFleet();
     }
 
-    public Boolean somethingWasHitted() {
+    public Boolean somethingWasHit(Map<Integer, Boolean> map) {
+    	for(Entry<Integer, Boolean> markIndex: map.entrySet()) {
```
Comment (mprtcz) [inline src/main/java/com/java_academy/logic/state_machine/PlayerEndActionState.java:55]: Take a look at 
`return map.entrySet().stream().anyMatch(Map.Entry::getValue);`

File: src/main/java/com/java_academy/logic/state_machine/core/GameState.java (hunk 6 above)
Comment (mprtcz) [inline src/main/java/com/java_academy/logic/state_machine/core/GameState.java:19]: 👍

File: src/main/java/com/java_academy/network/input/SocketInputDataProcessor.java [hunk 14]
```diff
@@ -55,11 +58,12 @@ public void run() {
                 messageReceived(input);
             }
 
-        } catch (EOFException e){
-            System.out.println("Socket closed! InputProcessor execution terminated!");
+        } catch (SocketException | EOFException ex){
+            System.out.println("Socket closed!");
```
Comment (mprtcz) [inline src/main/java/com/java_academy/network/input/SocketInputDataProcessor.java:62]: Not even a stack trace?

File: src/main/resources/Messages_en_EN.properties [hunk 15]
```diff
@@ -9,4 +9,11 @@ TRYING_TO_TERMINATE_EXECUTOR = Trying to terminate executor...
 CANT_SEND_MESSAGE_TO_PLAYER = Can\'t send a message to a player: %s, socket was closed!
 CANT_CREATE_STREAM = Can\'t create stream! Socket was closed!
 CLOSE_MESSAGE_RECEIVED = Close message received!
-CANT_SEND_MESSAGE_TO_SERVER = Cant send message to the server. Socket was closed!
\ No newline at end of file
+CANT_SEND_MESSAGE_TO_SERVER = Cant send message to the server. Socket was closed!
```
Comment (mprtcz) [inline src/main/resources/Messages_en_EN.properties:12]: i18n 👍

File: src/test/java/com/java_academy/logic/StateMachine/PlayerEndActionStateTest.java [hunk 16]
```diff
@@ -19,7 +19,7 @@
     @BeforeTest
     public void setUp(){
         player = Players.FIRST_PLAYER;
-        playerEndActionState = new PlayerEndActionState(player);
+        playerEndActionState = new PlayerEndActionState(player, null);
```
Comment (mprtcz) [inline src/test/java/com/java_academy/logic/StateMachine/PlayerEndActionStateTest.java:22]: Maybe some mock instead of a bare null?

**Review by mprtcz — COMMENTED** [review]
Some minor comments about the code, I suggest to look at the map stream code line and a lack of stack trace

File: src/main/java/com/java_academy/ClientApp.java (hunk 2 above)
Comment (arturAR) [reply src/main/java/com/java_academy/ClientApp.java:33]: This was just for demo. In next commits will be repaire.

File: src/main/java/com/java_academy/logic/state_machine/PlayerActionState.java (hunk 11 above)
Comment (arturAR) [reply src/main/java/com/java_academy/logic/state_machine/PlayerActionState.java:52]: This is for refactor, I remember :)

File: src/test/java/com/java_academy/logic/StateMachine/PlayerEndActionStateTest.java (hunk 16 above)
Comment (arturAR) [reply src/test/java/com/java_academy/logic/StateMachine/PlayerEndActionStateTest.java:22]: In this example it doesn't matter.

**Pull request merged**